File: core/modules/users/user.service.ts
/**
 * User Service
 */
import { IUserRepository, UserFilter, PaginatedUsers } from './user.repository.interface.js';
import { User, Role, Permission } from '../../../models/index.js';

let UserRepository: IUserRepository;
let AuthService: any;
let auditService: any;
let hashPassword: any;

export function initUserService(
    repo: IUserRepository,
    authSvc: any,
    audit: any,
    hashPwd: any
) {
    UserRepository = repo;
    AuthService = authSvc;
    auditService = audit;
    hashPassword = hashPwd;
}

interface CreateUserInput {
    username: string;
    name: string;
    password: string;
    role_id?: number;
    is_active?: number;
    createdBy: number;
}

interface UpdateUserInput {
    username?: string;
    name?: string;
    password?: string;
    role_id?: number;
    is_active?: number;
    updatedBy: number;
}

function normalizeUsername(username: string): string {
    return username.trim();
}

function ensureRoleExists(roleId: number): void {
    const roleExists = UserRepository.getAllRoles().some(role => role.id === roleId);
    if (!roleExists) {
        throw new Error('Invalid role selected');
    }
}

export async function getUsers(params: UserFilter): Promise<PaginatedUsers> {
    return UserRepository.findUsers(params);
}

export async function getUserById(id: number): Promise<User> {
    const user = UserRepository.findUserById(id);
    if (!user) throw new Error(`User ${id} not found`);
    return user;
}

export async function createUser(data: CreateUserInput): Promise<User> {
    const trimmedName = data.name.trim();
    if (!trimmedName) throw new Error('Name is required');

    const username = normalizeUsername(data.username);
    if (username.length < 3) throw new Error('Username must be at least 3 characters');
    const existing = UserRepository.findUserByUsername(username);
    if (existing) throw new Error('Username already exists');

    const password_hash = await hashPassword(data.password);
    if (data.role_id !== undefined) {
        ensureRoleExists(data.role_id);
    }
    const roleIds = data.role_id ? [data.role_id] : [];

    const newUser = UserRepository.insertUserWithRoles({
        name: trimmedName,
        username,
        password_hash
    }, roleIds);

    if (!newUser) throw new Error('Failed to create user');

    const userId = newUser.id;

    // Log user creation (exclude password)
    const { password, ...logData } = { ...data, id: userId, username };
    auditService.logCreate(data.createdBy!, 'users', userId, logData);

    return newUser;
}

export async function updateUser(id: number, data: UpdateUserInput): Promise<User> {
    const oldUser = await getUserById(id); // Ensure exists

    if (data.name !== undefined && !data.name.trim()) {
        throw new Error('Name is required');
    }
    if (data.name !== undefined) {
        data.name = data.name.trim();
    }

    if (data.username && data.username !== oldUser.username) {
        const normalizedUsername = normalizeUsername(data.username);
        if (normalizedUsername.length < 3) {
            throw new Error('Username must be at least 3 characters');
        }
        const existing = UserRepository.findUserByUsername(normalizedUsername);
        if (existing) throw new Error('Username already taken');
        data.username = normalizedUsername;
    }

    const { password, role_id, updatedBy, ...restData } = data;
    const updateData: Partial<User> = { ...restData };
    if (password) {
        updateData.password_hash = await hashPassword(password);
    }
    if (role_id !== undefined) {
        ensureRoleExists(role_id);
    }
    const roleIds = role_id !== undefined ? [role_id] : undefined;

    const updatedUser = UserRepository.updateUserWithRolesById(id, updateData, roleIds);
    if (!updatedUser) throw new Error('Failed to update user');

    // Enforce immediate permission/session consistency after account-sensitive updates.
    if (data.is_active === 0 || role_id !== undefined || !!password) {
        AuthService.revokeUserSessions(id);
    }

    // Log user update (exclude password)
    const { password_hash: _old, ...oldData } = oldUser;
    const { password_hash: _new, ...newData } = updatedUser;
    auditService.logUpdate(updatedBy!, 'users', id, oldData, newData);

    return updatedUser;
}

export async function deleteUser(id: number, deletedBy: number): Promise<boolean> {
    if (id === deletedBy) {
        throw new Error('You cannot delete your own account');
    }

    const user = await getUserById(id);
    const success = UserRepository.softDeleteUser(id);

    // Log user deletion (exclude password)
    if (success) {
        AuthService.revokeUserSessions(id);
        const { password_hash, ...userData } = user;
        auditService.logDelete(deletedBy, 'users', id, userData);
    }

    return success;
}

export async function getRoles(): Promise<Role[]> {
    return UserRepository.getAllRoles();
}

export async function getPermissions(): Promise<Permission[]> {
    return UserRepository.getAllPermissions();
}

export async function getUserRoles(userId: number): Promise<Role[]> {
    const user = await getUserById(userId);
    return UserRepository.getUserRoles(userId);
}

export async function updateUserRoles(userId: number, roleIds: number[]): Promise<void> {
    const user = await getUserById(userId);

    // Validate all role IDs exist
    const allRoles = UserRepository.getAllRoles();
    for (const roleId of roleIds) {
        if (!allRoles.some(r => r.id === roleId)) {
            throw new Error(`Invalid role ID: ${roleId}`);
        }
    }

    UserRepository.assignUserRoles(userId, roleIds);
    AuthService.revokeUserSessions(userId);
}

export async function getUserPermissionOverrides(userId: number): Promise<Array<{ permission_id: number; key: string; granted: number }>> {
    const user = await getUserById(userId);
    return UserRepository.getUserPermissionOverrides(userId);
}

export async function updateUserPermissions(userId: number, permissions: Array<{ permissionId: number; granted: boolean }>): Promise<void> {
    const user = await getUserById(userId);

    // Validate all permission IDs exist
    const allPermissions = UserRepository.getAllPermissions();
    for (const perm of permissions) {
        if (!allPermissions.some(p => p.id === perm.permissionId)) {
            throw new Error(`Invalid permission ID: ${perm.permissionId}`);
        }
    }

    // Apply each permission override
    for (const perm of permissions) {
        UserRepository.setUserPermission(userId, perm.permissionId, perm.granted);
    }

    AuthService.revokeUserSessions(userId);
}

File: src/main/java/com/possum/application/people/UserService.java
package com.possum.application.people;

import com.possum.domain.model.Permission;
import com.possum.domain.model.Role;
import com.possum.domain.model.User;
import com.possum.domain.model.UserPermissionOverride;
import com.possum.infrastructure.security.PasswordHasher;
import com.possum.persistence.repositories.interfaces.UserRepository;
import com.possum.shared.dto.PagedResult;
import com.possum.shared.dto.UserFilter;
import com.possum.shared.util.DomainValidators;

import com.possum.shared.util.TimeUtil;
import java.util.List;
import java.util.Optional;

public class UserService {
    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;

    public UserService(UserRepository userRepository, PasswordHasher passwordHasher) {
        this.userRepository = userRepository;
        this.passwordHasher = passwordHasher;
    }

    public PagedResult<User> getUsers(UserFilter filter) {
        return userRepository.findUsers(filter);
    }

    public Optional<User> getUserById(long id) {
        return userRepository.findUserById(id);
    }

    public User createUser(String name, String username, String password, boolean active, List<Long> roleIds) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.USERS_MANAGE);
        if (name == null || name.isBlank()) throw new com.possum.domain.exceptions.ValidationException("User name is required");
        if (username == null || username.isBlank()) throw new com.possum.domain.exceptions.ValidationException("Username is required");
        if (username.contains(" ")) throw new com.possum.domain.exceptions.ValidationException("Username cannot contain spaces");
        if (username.length() < DomainValidators.MIN_USERNAME_LENGTH) throw new com.possum.domain.exceptions.ValidationException("Username must be at least " + DomainValidators.MIN_USERNAME_LENGTH + " characters");
        if (password == null || password.isBlank()) throw new com.possum.domain.exceptions.ValidationException("Password is required");
        if (password.length() < DomainValidators.MIN_PASSWORD_LENGTH) throw new com.possum.domain.exceptions.ValidationException("Password must be at least " + DomainValidators.MIN_PASSWORD_LENGTH + " characters");
        if (userRepository.findUserByUsername(username.trim()).isPresent())
            throw new com.possum.domain.exceptions.ValidationException("Username '" + username.trim() + "' is already taken");
        
        String hashedPassword = passwordHasher.hashPassword(password);
        User newUser = new User(null, name, username, hashedPassword, active, TimeUtil.nowUTC(), TimeUtil.nowUTC(), null);
        return userRepository.insertUserWithRoles(newUser, roleIds);
    }

    public User updateUser(long id, String name, String username, String password, boolean active, List<Long> roleIds) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.USERS_MANAGE);
        if (name == null || name.isBlank()) throw new com.possum.domain.exceptions.ValidationException("User name is required");
        if (username == null || username.isBlank()) throw new com.possum.domain.exceptions.ValidationException("Username is required");
        User existingUser = userRepository.findUserById(id)
                .orElseThrow(() -> new com.possum.domain.exceptions.NotFoundException("User not found: " + id));
        
        String hashedPassword = existingUser.passwordHash();
        if (password != null && !password.trim().isEmpty()) {
            hashedPassword = passwordHasher.hashPassword(password);
        }

        User updatedUser = new User(existingUser.id(), name, username, hashedPassword, active, existingUser.createdAt(), TimeUtil.nowUTC(), existingUser.deletedAt());
        User result = userRepository.updateUserWithRolesById(id, updatedUser, roleIds);
        
        if (!active) {
            userRepository.revokeUserSessions(id);
        }
        return result;
    }

    public void deleteUser(long id) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.USERS_MANAGE);
        if (!userRepository.softDeleteUser(id)) {
            throw new com.possum.domain.exceptions.NotFoundException("User not found: " + id);
        }
    }

    public List<Role> getAllRoles() {
        return userRepository.getAllRoles();
    }

    public List<Permission> getAllPermissions() {
        return userRepository.getAllPermissions();
    }

    public List<Role> getUserRoles(long userId) {
        return userRepository.getUserRoles(userId);
    }

    public List<String> getUserPermissions(long userId) {
        return userRepository.getUserPermissions(userId);
    }

    public void assignUserRoles(long userId, List<Long> roleIds) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.USERS_MANAGE);
        userRepository.assignUserRoles(userId, roleIds);
    }

    public List<UserPermissionOverride> getUserPermissionOverrides(long userId) {
        return userRepository.getUserPermissionOverrides(userId);
    }

    public void setUserPermission(long userId, long permissionId, boolean granted) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.USERS_MANAGE);
        userRepository.setUserPermission(userId, permissionId, granted);
    }

    public List<Long> getRolePermissions(List<Long> roleIds) {
        return userRepository.getRolePermissions(roleIds);
    }
}
